### PR TITLE
HDDS-11694. Safemode Improvement: Introduce factory class to create safemode rules.

### PR DESCRIPTION
**Pull request by nandakumar131**
## What changes were proposed in this pull request?
Safemode Improvement: Introduce factory class to create safemode rules.

Currently, the SafeModeRule creation is tightly coupled within SafeModeManager. This can be decoupled with the introduction of Factory Class for Rule creation.

This will help us in implementing annotation based rule loading in future.

## What is the link to the Apache JIRA
HDDS-11694

## How was this patch tested?
Additional unit tests added to verify the Factory class behaviour.

### CODE REVIEW

**Comment by adoroszlai** [comment]
Thanks @nandakumar131 for the patch.  Please enable the [build-branch](https://github.com/nandakumar131/ozone/actions/workflows/post-commit.yml) workflow in your fork.

**Comment by nandakumar131** [comment]
@adoroszlai, I have enabled [build-branch](https://github.com/nandakumar131/ozone/actions/workflows/post-commit.yml) workflow in my fork.

**Comment by errose28** [comment]
Looks like there's still pending TODOs in the code. Should this be a draft PR?

**Comment by nandakumar131** [comment]
Thanks @errose28 for taking a look at this.

The motivation here is to simplify and modularize the `SafeMode` code in SCM, as it got complex and messy over time.

As I started making changes, I realised that it will be hard for someone to review the whole change in a single PR.
So, I'm breaking it into multiple (easy to understand and review) Jiras/PRs. The whole task is tracked under the parent Jira HDDS-11693.

The TODO is added here to give hint to the reviewer that this code will be optimized/modified in the follow-up Jiras/PRs.

**Comment by errose28** [comment]
Got it. Could you add Jira IDs to the TODOs so it is easier to track? Untracked TODO messages tend to get missed and become permanent parts of the code base 😄

**Comment by nandakumar131** [comment]
@errose28 , updated the Jira IDs to the TODOs.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/safemode/TestSafeModeRuleFactory.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+package org.apache.hadoop.hdds.scm.safemode;
+
+import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.hdds.scm.container.ContainerManager;
+import org.apache.hadoop.hdds.scm.ha.SCMContext;
+import org.apache.hadoop.hdds.scm.pipeline.PipelineManager;
+import org.apache.hadoop.hdds.server.events.EventQueue;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
```
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/safemode/TestSafeModeRuleFactory.java:27]: nit: please use static import for assertions and mocks (see HDDS-9961).

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/SafeModeRuleFactory.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+          ATLEAST_ONE_DATANODE_REPORTED_PIPELINE_EXIT_RULE, eventQueue,
+          pipelineManager, safeModeManager, config));
+    }
+
+  }
+
+  public static SafeModeRuleFactory getInstance() {
+    if (instance != null) {
+      return instance;
+    }
```
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/SafeModeRuleFactory.java:107]: For thread-safety,
- `getInstance()` should be `synchronized`, OR
- `instance` should be `volatile`, and assigned to local variable first

**Review by adoroszlai — APPROVED** [review]
Thanks @nandakumar131 for the improvement.  My comments can be addressed in one of the follow-up PRs under HDDS-11693.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/SafeModeRuleFactory.java (hunk 2 above)
Comment (ChenSammi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/SafeModeRuleFactory.java:107]: Same suggestion.  

Except this, others looks good to me.

Comment (nandakumar131) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/SafeModeRuleFactory.java:107]: Not sure how I missed adding `synchronized` here. Thanks for the catch @adoroszlai  & @ChenSammi.

**Comment by adoroszlai** [comment]
Thanks @nandakumar131 for the patch, @ChenSammi, @errose28 for the review.

**Comment by nandakumar131** [comment]
Thanks @adoroszlai, @ChenSammi & @errose28!

**Pull request merged**